synapseclient/client.py

```diff
@@ -4823,7 +4824,7 @@ def getSubmission(self, id, **kwargs):
              on the *downloadFile*, *downloadLocation*, and *ifcollision* parameters
         """
 
-        submission_id = id_of(id)
+        submission_id = validate_submission_id(id)
```

Comment:
Just making note that this should still work:

```
import synapseclient
syn = synapseclient.login()
syn.getSubmission(<id here>)
sub_obj = syn.getSubmission(<id here>)
syn.getSubmission(sub_obj)
```

Reply:
This is a perfect example of why type hints would've been great when this function was originally implemented 😅 
Furthermore, there _should be_ an integration test for this scenario.

Reply:
Hmm, why would a person get a submission object with the submission object itself, if the submission object is what they're trying to retrieve? Hope that makes sense

Reply:
> My recommendation is to not add complexity to the validate_submission_id function to check for the Submission object,

I can see this pov and was thinking on those same lines, but I started fiddling with the code (see changes [here](https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1135/commits/c33f42448868ac9f88439f2745e72ca9d8be8ff0)) and I don't think it adds much more complexity:

This added an extra conditional and introduced the same dictionary-handling function that's in `id_of`, which makes `validate_submission_id` have the same level of abstraction as `id_of`. From my pov, it makes `id_of` and `validate_submission_id` look more like analog functions, which is what we're going for here because that's essentially what they are.

```
if isinstance(submission_id, int):
    return str(submission_id)
elif isinstance(submission_id, str) and submission_id.isdigit():
    return submission_id
elif isinstance(submission_id, collections.abc.Mapping):
    syn_id = _get_from_members_items_or_properties(submission_id, 'id')
    if syn_id is not None:
        return str((float(syn_id)))
```

With these changes, we now have support for input Submission objects in `getSubmission` and `getSubmissionStatus`:

```
In [18]: sub = syn.getSubmissionStatus("9745366.0")
[WARNING] Submission ID '9745366.0' contains decimals which are not supported. Submission ID will be converted to '9745366'.

In [19]: sub2 = syn.getSubmissionStatus(sub)

In [20]: sub3 = syn.getSubmission(sub2)

In [21]: sub3
Out[21]: 
{'id': '9745366',
...
```

I haven't added tests for this because we're still deciding. but if we are keeping support for `getSubmissionBundles` and Submission objects as inputs for `getSubmission` and `getSubmissionStatus`, then I'm voting for keeping `validate_submission_id` in the state that it's in in https://github.com/Sage-Bionetworks/synapsePythonClient/pull/1135/commits/c33f42448868ac9f88439f2745e72ca9d8be8ff0

Reply:
> With these changes, we now have support for input Submission objects in getSubmission and getSubmissionStatus:

This is great, and what I would expect as of right now in the client.

Some thoughts on future state:

I would like for this to eventually evolve to work similarly to the function signature I've written for the `entity_factory.py` defined here: https://github.com/Sage-Bionetworks/synapsePythonClient/blob/develop/synapseclient/api/entity_factory.py#L26-L38 . The idea is that any `client/Synapse` class methods are only used as a factory for these classes, but then any operations take place of those objects themselves. For example you'd call `.get_submission()` or `.get_submissions()` and it's return a `Submission` object which would have a `.status` attribute on the class instance.

Reply:
Thanks for the feedback all - I moved forward with this solution and am working on one last unit test that ensures the inputs of `getSubmissionStatus` and `getSubmission` work together, and then this will be good to go for a final review.

Also noticed we don't have much code coverage for `getSubmissionStatus` itself, so I made a [tech debt ticket](https://sagebionetworks.jira.com/browse/IBCDPE-1093) to cover this.